src/backend/distributed/utils/shardinterval_utils.c

```diff
@@ -0,0 +1,293 @@
+/*-------------------------------------------------------------------------
+ *
+ * shardinterval_utils.c
+ *
+ * This file contains functions to perform useful operations on shard intervals.
+ *
+ * Copyright (c) 2014-2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "postgres.h"
+
+#include "access/nbtree.h"
+#include "catalog/pg_am.h"
+#include "catalog/pg_collation.h"
+#include "catalog/pg_type.h"
+#include "distributed/shardinterval_utils.h"
+#include "distributed/pg_dist_partition.h"
+#include "distributed/worker_protocol.h"
+#include "utils/catcache.h"
+#include "utils/memutils.h"
+
+
+/*
+ * ShardIntervalCompareFunctionCache is used for caching shard interval
+ * compare functions.
+ */
+static List *ShardIntervalCompareFunctionCache = NIL;
+
+
+static FmgrInfo * ShardIntervalCompareFunction(Var *partitionColumn,
+											   char partitionMethod);
+static ShardInterval * SearchCachedShardInterval(Datum partitionColumnValue,
+												 ShardInterval **shardIntervalCache,
+												 int shardCount,
+												 FmgrInfo *compareFunction);
+
+
+/*
+ * CompareShardIntervals acts as a helper function to compare two shard intervals
+ * by their minimum values, using the value's type comparison function.
+ *
+ * If a shard interval does not have min/max value, it's treated as being greater
+ * than the other.
+ */
+int
+CompareShardIntervals(const void *leftElement, const void *rightElement,
+					  FmgrInfo *typeCompareFunction)
+{
+	ShardInterval **leftShardInterval = (ShardInterval **) leftElement;
+	ShardInterval **rightShardInterval = (ShardInterval **) rightElement;
+	Datum leftDatum = 0;
+	Datum rightDatum = 0;
+	int comparisonResult = 0;
+
+	/*
+	 * Left element should be treated as the greater element in case it doesn't
+	 * have min or max values.
+	 */
+	if (!(*leftShardInterval)->minValueExists || !(*leftShardInterval)->maxValueExists)
+	{
+		comparisonResult = 1;
+		return comparisonResult;
+	}
+
+	/*
+	 * Right element should be treated as the greater element in case it doesn't
+	 * have min or max values.
+	 */
+	if (!(*rightShardInterval)->minValueExists || !(*rightShardInterval)->maxValueExists)
+	{
+		comparisonResult = -1;
+		return comparisonResult;
+	}
+
+	/* if both shard interval have min/max values, calculate the comparison result */
+	leftDatum = (*leftShardInterval)->minValue;
+	rightDatum = (*rightShardInterval)->minValue;
+
+	Datum comparisonDatum = CompareCall2(typeCompareFunction, leftDatum, rightDatum);
```

Comment:
This will warn in some compilers, because you're declaring a variable after code.